### PR TITLE
Beta

### PR DESCRIPTION
**Pull request by lucasarlop**
- home page (beta.askbywho.com/): shows a form with brands, name and email;
    - if the email already exists, a update is done;
- statistics page (beta.askbywho.com/:id/share): showed after filling the form on home page;

on admin:
- /users
- /brands
- /emails

others:
* tests
* validations
* mobile layout

### CODE REVIEW

File: priv/repo/migrations/20170710175356_create_email.exs [hunk 1]
```diff
@@ -0,0 +1,20 @@
+defmodule Askbywho.Repo.Migrations.CreateEmail do
```
Comment (ohanhi) [inline priv/repo/migrations/20170710175356_create_email.exs:1]: It seems to me the current database schema has these properties:
- `email` has to be unique (a site visitor can't submit multiple times using the same email) 
- brands `name` has to be unique (the same name can only be added by the first user submitting it)
- the connecting table doesn't have timestamps

I have two questions about this:
1. Is it really useful to have the unique constraint on the emails?
2. Do we want to answer the question _when did Anna submit ASOS_? If we remove the unique constraint on the emails, we can't really tell (for now, the `created_at` of the email is the moment when all submissions from that email had to happen).

File: test/controllers/page_controller_test.exs [hunk 2]
```diff
@@ -1,8 +1,23 @@
...
+
+  @valid_attrs %{email: "some@email.com", name: "Some Content"}
+  @invalid_attrs %{}
+
+  test "renders form for new subscriber", %{conn: conn} do
+    conn = get conn, page_path(conn, :index)
+    assert html_response(conn, 200) =~ "Ssh! Welcome to our secret beta!"
   end
+
+  # test "create new subscriber and redirects when data is valid", %{conn: conn} do
```
Comment (ohanhi) [inline test/controllers/page_controller_test.exs:14]: Either make the test useful or remove the commented-out code.

File: web/controllers/email_controller.ex [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+      {:ok, _email}       -> # Inserted or updated with success
+        conn
+        |> put_flash(:info, "Email created successfully.")
+        |> redirect(to: email_path(conn, :index))
+      {:error, changeset} -> # Something went wrong
+        render(conn, "new.html", changeset: changeset)
+    end
+  end
+
+#  def create(conn, %{"email" => email_params}) do
```
Comment (ohanhi) [inline web/controllers/email_controller.ex:37]: Remove commented-out code :)

File: web/templates/layout/site.html.eex [hunk 4]
```diff
@@ -0,0 +1,19 @@
+<!DOCTYPE html>
+<html lang="en">
+  <head>
+    <meta charset="utf-8">
+    <meta http-equiv="X-UA-Compatible" content="IE=edge">
+    <meta name="viewport" content="width=device-width, initial-scale=1">
+    <meta name="description" content="">
```
Comment (ohanhi) [inline web/templates/layout/site.html.eex:7]: @ilonamm We probably want a proper description here.

Comment (ohanhi) [reply web/templates/layout/site.html.eex:7]: Maybe also the `og:` tags we talked about earlier.

File: web/templates/page/index.html.eex [hunk 5]
```diff
@@ -1,36 +1,85 @@
...
+<div id="home-wrapper">
+  <div id="beta-message">
+    <div class="container">
+      <h1>
+        Ssh! Welcome to our secret beta!
+      </h1>
+
+      <p>
+        Thanks to you we can launch a site that doesn’t look deserted. Please
+        nominate at least five clothing brands, and leave your name & email so
```
Comment (ohanhi) [inline web/templates/page/index.html.eex:10]: Use `&amp;` for the ampersand.

File: web/templates/page/share.html.eex [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+  <div class="foot">
+    <div class="container">
+      <p>
+        Follow us
+      </p>
+      
+      <div class="social">
+        <a href="https://www.facebook.com/sharer/sharer.php?u=www.askbywho.com">
+          <img src="/images/facebook.png">
```
Comment (ohanhi) [inline web/templates/page/share.html.eex:77]: Add `alt="Share on Facebook"` or similar for the Facebook and Twitter images.

File: web/templates/page/share.html.eex [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+        </a>
+
+        <a href="https://twitter.com/home?status=www.askbywho.com">
```
Comment (ohanhi) [inline web/templates/page/share.html.eex:80]: @ilonamm Let's come up with a nice base tweet instead of just the URL.

**Review by ohanhi — COMMENTED** [review]
I added some comments about the database schema, plus minor notes about the HTML.

I think we should also decide on a solid CSS strategy before we accumulate a lot of styles. Here's a nice article about BEM, for example: https://csswizardry.com/2013/01/mindbemding-getting-your-head-round-bem-syntax/

**Comment by batmany13** [comment]
FYI, I'm going manually merge in the changes as there were many changes to mix.exs

**Comment by lucasmncastro** [comment]
@ohanhi Great review! Very detailed and important notices.

About CSS, thank you so much for the tips, I really didn't know the BEM Strategy, it seems interesting!

Let's improve this in the next versions.

File: web/templates/page/share.html.eex (hunk 7 above)
Comment (ilonamm) [reply web/templates/page/share.html.eex:80]: @ohanhi Yes! Think I have that on trello.

File: config/config.exs [hunk 8]
```diff
@@ -30,3 +30,20 @@ config :phoenix, :generators,
...
+
+# %% Coherence Configuration %%   Don't remove this line
+config :coherence,
+  user_schema: Askbywho.User,
+  repo: Askbywho.Repo,
+  module: Askbywho,
+  router: Askbywho.Router,
+  messages_backend: Askbywho.Coherence.Messages,
+  logged_out_url: "/",
+  email_from_name: "Your Name",
```
Comment (batmany13) [inline config/config.exs:42]: We probably change this to a valid user when we launch

File: config/dogma.exs [hunk 9]
```diff
@@ -2,10 +2,66 @@
...
+      %Rule.WindowsLineEndings{},
+    ]
+  end
+end
+
 config :dogma,
 
   # Select a set of rules as a base
-  rule_set: Dogma.RuleSet.All,
+  rule_set: AbwRuleSet,
```
Comment (batmany13) [inline config/dogma.exs:63]: +1

File: priv/repo/migrations/20170710175356_create_email.exs (hunk 1 above)
Comment (batmany13) [reply priv/repo/migrations/20170710175356_create_email.exs:1]: Mmm, the uniqueness of emails would make sense so you don't have multiple entries for the same email address, however, if we added the timestamps to the linking table, then if the same user submitted multiple times, we'd be able to see the submissions and the time of those submissions.  Thus, the "brands_emails" table really is a submissions history, and then the "emails" and "brands" table is really a unique list of names.

File: web/models/brand.ex [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+  @doc """
+  Creates a changeset based on the `model` and `params`.
+
+  If no params are provided, an invalid changeset is returned
+  with no validation performed.
+  """
+  def changeset(struct, params \\ %{}) do
+    struct
+    |> cast(params, [:name])
+    |> validate_required([:name])
```
Comment (batmany13) [inline web/models/brand.ex:20]: What do we do about mix cased brands?  So what if some typed "Hugo Boss" vs "hugo boss"?  Also, we may want to use something like [jaro](https://hexdocs.pm/elixir/String.html#jaro_distance/2) distance to determine if names are similar (found usually when you're about 0.90, its most likely the same name).  That can come later :)

File: web/models/email.ex [hunk 11]
```diff
@@ -0,0 +1,50 @@
+defmodule Askbywho.Email do
+  use Askbywho.Web, :model
+  alias Askbywho.Repo
+  alias Askbywho.Brand
+
+  schema "emails" do
```
Comment (batmany13) [inline web/models/email.ex:6]: so at what point would an "email' be turned to a coherence user?  So if an user does a quick signup, shouldn't they just be a coherence user who hasn't created a password?  Thus, couldn't we just use the user table as the links to the brands created?  Then, we can try to convert users who don't have account information to add more detail?  Also, what does logging in buy us?  Maybe this is covered later in the code :)

File: web/views/page_view.ex [hunk 12]
```diff
@@ -4,4 +4,54 @@ defmodule Askbywho.PageView do
...
+    "#{count} brands"
+  end
+
+  def count_brands do
+    Repo.aggregate Brand, :count, :id
+  end
+
+  def count_nominations do
+    {_status, result} = Ecto.Adapters.SQL.query(
+      Repo, "select count(1) from brands_emails"
```
Comment (batmany13) [inline web/views/page_view.ex:31]: We typically shouldn't be putting Repo logic in the "views".  Typically the controller should load this information, and all calculation logic should be put in the Models themselves.  So, things like counts and percentage should all be gotten from the Repo, and then loaded as vars in the controller itself

File: test/views/page_view_test.exs [hunk 13]
```diff
@@ -1,3 +1,86 @@
 defmodule Askbywho.PageViewTest do
   use Askbywho.ConnCase, async: true
+  alias Askbywho.PageView
+  alias Askbywho.Repo
+  alias Askbywho.Email
+  alias Askbywho.Brand
+
+  test "#count_emails should get the number of emails on database" do
+    %Email{}
+      |> Email.changeset(%{"email" => "test@email.com", "name" => "Test", "brands" => "Adidas, Nike"})
```
Comment (batmany13) [inline test/views/page_view_test.exs:10]: Should have "DRY" here, and consider creating a seed function that would just pre-popular the database with the right base set of emails.

File: test/views/page_view_test.exs [hunk 14]
```diff
@@ -1,3 +1,86 @@
...
+      |> Repo.insert
+
+    %Email{}
+      |> Email.changeset(%{"email" => "test2@email.com", "name" => "Test 2", "brands" => "Adidas, Cavalera"})
+      |> Repo.insert
+
+    adidas = Repo.get_by(Brand, %{name: "Adidas"})
+
+    assert PageView.percent_nominations(adidas.id) == 0.002
```
Comment (batmany13) [inline test/views/page_view_test.exs:57]: This is a perfect example of having the logic in the Brand object so you can just call Brand.percent_nominations(adidas.id)

**Review by batmany13 — APPROVED** [review]
Great job on your first "batch commit".  A couple of comments/suggestions for the future

1.  Try to break up your commits into smaller logical chunks.  So for instance, you could create the data model first, and add that as an initial PR.  Then, add the base logic for brands/emails.  Finally, adding something like coherence would've been a great final PR, that way, you would've been able to introduce changes in bite sized chunks, and also had eyes on it.  

Of course you guys did this in blazing time so not too much harm :)

2.  This may be a convention we just decide on, but the view of doing "per feature" branching vs group branching, it seems like this worked out pretty well, so don't want to ruin the flow, but I think as our code gets more complex, we'll probably want to start creating branches for each specific feature

**Comment by batmany13** [comment]
I did actually have one last comment, how do you think we should handle suggestions and track them?  I don't think we need to fix all PRs comments all at once and its good to track them and fix them incrementally, so just curious how you guys want to handle that?

**Comment by lucasmncastro** [comment]
Great code review! We are learning a lot with this comments, thank you guys!

It was really a huge PR with a lot of things to improve; we wanted to put it on air in order as soon as possible in order to help us to validate some business hypothesis, so breaking it down in smaller branches seemed a little bureaucratic at this stage.

About those technical suggestions, let's prioritize them?

**Comment by lucasmncastro** [comment]
Other question: do you have some suggestion about the branches names?

File: web/templates/layout/site.html.eex (hunk 4 above)
Comment (ilonamm) [reply web/templates/layout/site.html.eex:7]: Added this in Trello on this card:
https://trello.com/c/phqVVjAQ/10-set-up-fb-insta-twitter-sharing-content

File: web/templates/page/share.html.eex (hunk 7 above)
Comment (ilonamm) [reply web/templates/page/share.html.eex:80]: This is in Trello as https://trello.com/c/phqVVjAQ/10-set-up-fb-insta-twitter-sharing-content

File: priv/repo/migrations/20170710175356_create_email.exs (hunk 1 above)
Comment (ilonamm) [reply priv/repo/migrations/20170710175356_create_email.exs:1]: Unique email makes sense from the user perspective, I added a reminder to test how this works and affects statistics on https://trello.com/c/H3Nt9wb9/21-qa-full-site-testing

Unique email testing:
If a lisa@email.com submits Adidas, Puma and Nike today, and returns tomorrow to submit Adidas, Zara and Mango, lisa@email.com should associate Adidas, Puma, Nike, Zara, and Mango. 

In our statistics Adidas should be counted once, and Lisa@email.com should be counted as one person.

File: web/models/email.ex (hunk 11 above)
Comment (ilonamm) [reply web/models/email.ex:6]: Added a trello card on this https://trello.com/c/oV3auVsy/35-plan-user-management

File: config/config.exs (hunk 8 above)
Comment (ilonamm) [reply config/config.exs:42]: Added this in Trello https://trello.com/c/oV3auVsy/35-plan-user-management

File: web/models/brand.ex (hunk 10 above)
Comment (ilonamm) [reply web/models/brand.ex:20]: Made a Trello card on this: https://trello.com/c/eoX1xmaN/30-figure-out-how-to-manage-and-avoid-misspelled-brand-names

File: web/views/page_view.ex (hunk 12 above)
Comment (ilonamm) [reply web/views/page_view.ex:31]: @batmany13 could you add here if this is a must-fix, proposal, or feedback?

File: test/views/page_view_test.exs (hunk 13 above)
Comment (ilonamm) [reply test/views/page_view_test.exs:10]: @batmany13 could you add here if this is a must-fix, proposal, or feedback?

File: test/views/page_view_test.exs (hunk 14 above)
Comment (ilonamm) [reply test/views/page_view_test.exs:57]: @batmany13 could you add here if this is a must-fix, proposal, or feedback?

**Comment by ilonamm** [comment]
@lucasmncastro Added a simple guide on branch naming to our readme: "When naming branches, use yourname-featurename, for example ilona-herokusetup. When changes have been merged, delete the branch." :)

**Pull request merged**